exonum/src/runtime/mod.rs

```diff
@@ -432,15 +438,63 @@ impl<'a> ExecutionContext<'a> {
     /// Maximum depth of the call stack.
     const MAX_CALL_STACK_DEPTH: usize = 256;
 
-    pub(crate) fn new(dispatcher: &'a DispatcherRef<'a>, fork: &'a Fork, caller: Caller) -> Self {
+    pub(crate) fn new(dispatcher: &'a mut Dispatcher, fork: &'a mut Fork, caller: Caller) -> Self {
         Self {
+            dispatcher,
             fork,
             caller,
-            dispatcher,
             interface_name: "",
             call_stack_depth: 0,
         }
     }
+
+    pub(crate) fn child_context(&mut self, caller_service_id: InstanceId) -> ExecutionContext {
+        ExecutionContext {
+            dispatcher: &mut *self.dispatcher,
+            fork: self.fork,
+            caller: Caller::Service {
+                instance_id: caller_service_id,
+            },
+            interface_name: "",
+            call_stack_depth: self.call_stack_depth + 1,
+        }
+    }
+
+    pub(crate) fn call(
+        &mut self,
+        interface_name: &str,
+        call_info: &CallInfo,
+        arguments: &[u8],
+    ) -> Result<(), ExecutionError> {
+        if self.call_stack_depth >= ExecutionContext::MAX_CALL_STACK_DEPTH {
+            let kind = crate::runtime::dispatcher::Error::StackOverflow;
+            let msg = format!(
+                "Maximum depth of call stack has been reached. `MAX_CALL_STACK_DEPTH` is {}.",
+                ExecutionContext::MAX_CALL_STACK_DEPTH
+            );
+            return Err((kind, msg).into());
+        }
+
+        let runtime = self.dispatcher.runtime_for_service(call_info.instance_id)?;
+        let reborrowed = ExecutionContext {
+            fork: &mut *self.fork,
+            caller: self.caller,
+            interface_name,
+            dispatcher: &mut *self.dispatcher,
+            call_stack_depth: self.call_stack_depth,
```

Comment:
Shouldn't `call_stack_depth` be increased?

Reply:
I think not; this code is a copy-paste of the `reborrow()` method (with the exception of overriden `interface_name`), which is sort of like `&mut *r` for `r: &mut T`, only for owned data. In other words, the only purpose of this code is to mutably reborrow the inner contents of the context (i.e., `fork` and `dispatcher`) for some time.